### ISSUE TITLE
add (toggleable) support for more than 32 channels without the need to recompile from source

### ISSUE THREAD

**Issue by FyiurAmron** [milkytracker/MilkyTracker#62]
Preamble: I'm aware of the reasons why MT supports only 32 channels out-of-the-box ("real" FT2-ish XMs supporting only that much, MT being used on low-end platforms without CPU power to mix more channels, the "don't just double the channel amounts by 2" doctrine from FT2, already having the ability to use more channels with quite simple source hacks etc.) Still, I'd like to propose adding a support for more channels (at least 64) in master branch.

My points are:
a) more than 32 channels have been and already used by many (if not most) of the other contemporary trackers & players, including, but not limited to, IT, MPT, Skale, XMPlay etc., since increasing the amount of channels in XM-stored modules to e.g. 64 doesn't really require any other code changes than upping the constants in the application, so doesn't have any real cost other than making the XM incompatible with legacy software, and can (and does) provide real benefits (see below),
b) 32 channels for modules is simply too little for format without multilayered instruments or stereo samples; it happens for me regardless of the genre, both for e.g. orchestral music & chiptunes - it takes about 4 layers for a fat synth or a fat drum - double or triple that to 8 or 12 to make it possible to use stereo effects. I hit 32 channels before I'm even able to lay out the synths. It takes about 16 channels to have a properly sounding piano emulating two hands with chords played, 6 channels to emulate a guitar etc. ... so that's a real blocker. 64 channels usually leave you with some spare place, but 32 channels often leave you with not enough space for real layering and with the need to stack couple of instruments on single channel, killing the reverb/echo/ambience possibilities,
c) MT works properly (including GUI) recompiled with constants upped to 64/128 channels, at least on Win32 as far as the couple of weeks of testing I've done went, so there's no real effort needed here,
d) FT2 did that mostly because 32 channels with 48k sampling, effects and live playback was already pushing the hardware available at the time to the limit, and because 32 channels was an increase from FT1 and MOD format anyway; today, even an ARM smartphone has more CPU power than needed to mix 64 (or more) channels properly,
e) switching to 64/128 max could've been done with platform-dependent compilation, so that it only happens for those platforms where it's completely not an issue,
f) it could be/should be made an option (with a 1st time warning popping possibly), that XMs done in more-than-32-channels-mode are incompatible with legacy XM players/trackers,
g) I (and probably other people in similar situation to me) wouldn't have to maintain a fork and wouldn't have to port changes from master constantly just for the sake of having that particular feature.

IMO there would be no harm done, no kittens killed and no rainforests destroyed in the process. I can volunteer to do the PR with relevant changes myself, if nobody else wants to patch it together BTW.

**tl;dr** _please give us the ability to enable 64 (or more) channels out-of-the-box, so that the community gets the most of MT, which is awesome already, but can be even more awesome by having a killer feature other trackers already have for like 10 years._

**Comment by alugarius** [comment]
I'd like to see that to

**Comment by Deltafire** [comment]
I have no objection to this. I would suggest that a warning box is displayed when initially increasing the channels past 32, so the user knows they are creating an .XM file that is incompatible with FT2.

**Comment by SyntaxErrol** [comment]
I wonder if FT2 would still load and play the 32 first channels, and if so, would it kill the rest on save. I know it doesn't discard any pattern data when you for example reduce a 8 channel file to 4 and save. The channels can be restored on reload simply by upping the count again.

**Comment by alugarius** [comment]
@Deltafire I agree, if someone wants to use FT2, there should be a notification.

@SyntaxErrol I could try to get FT2 in a VM on my Debian machine and test it. I don't know how FT2 handles the data. If compatibility breaks, we have to save <= 32 channels files as before, and >32channel files in some sort of different format (we could leave the xm format as it is and the mod file format can be >32 ).
Thanks for comeing BTW :)

**Comment by alugarius** [comment]
I tested some music created with milkyTracker on Fastracker2, and some files are already incompatible. They are played on every tracker/player normally, but fastracker plays some files in "slowmotion"

**Comment by SyntaxErrol** [comment]
Just tested: Neither FT2 nor MT can open a 33-channel .XM created in OpenMPT citing "Corrupt file" and "Error while loading/unknown format" respectively.

> If compatibility breaks, we have to save <= 32 channels files as before, and >32channel files in some sort of different format (we could leave the xm format as it is and the mod file format can be >32 ).

This should probably follow OpenMPT, or is there another gold standard for nonstandard .XMs?

> We also could create a whole new format to seperate MT from FT2, opening an issue to a new file format and what name it should have. For example : extension = .mt or .mlky

Despite how much I love the Milky UI (nobody else seems to get the keyboard mapping "right" despite having "FT2 like" configurations) and having championed a hypothetical ".X3M" format in the past, it's hard for me to see a thing like this fit in the scope of the project. There was .SKM, there was .RNS which lives on as .XRNS and OpenMPT did the right thing to depart from .IT into .MPTM for their enhancements and the list goes on. We already have several more advanced trackers in existence so what could a MilkyTracker born format bring to the table that's different?

> some files are already incompatible. They are played on every tracker/player normally, but fastracker plays some files in "slowmotion"

That sounds strange. Did you try another setup (DOSBox maybe) or rendering to wav to rule out any realtime/VM specific shenanigans?

**Comment by alugarius** [comment]
" We already have several more advanced trackers in existence so what could a MilkyTracker born format bring to the table that's different? "

In the view of other trackers there is not so much difference, but in the view of milkytracker: beeing able to open impulse tracker files would be cool.

...How about using a format like IT to save a track up to 99 channels? No new format, just a new saving option?

**Comment by SyntaxErrol** [comment]
> Also OpenMPT is windows only.

Wow, somehow I thought that wasn't the case these days.

IT features with a familiar UI was always a common wish among FT2 users. Doubling the channel cap to match IT's would only be the start of supporting them. There are no virtual channels, no New Note Actions and no filters. And I don't think adding them is a matter of just adding code but changing a lot of what is already there possibly sacrificing FT2 compatibility - one of the project's main aspirations - in the process. That is without discrete feature/playback modes for different formats like there currently are for .XM and .MOD.

As I understand the FT2 and IT paradigms are somewhat at odds neither one being a superset of the other. Take instruments instruments for example: FT2's contain a set of samples sharing the same instrument settings and IT being able to define several instruments utilizing a single instance of sample data.

> MilkyTracker claims to beeing able to open Impulse tracker files

Actually .IT is just one on a [list of many formats](http://milkytracker.titandemo.org/docs/MilkyTracker.html#modules) that MilkyTracker is able to **import,** but is clear about 

> since Milky is a FT2 clone, modules are replayed in an FT2 environment which means not all features of different formats are supported.

Importing (and exporting) are considered lossy operations so for example while Milky's .S3M import and even playback is pretty good, features like channel panning and AdLib instruments just aren't supported. Bummer, especially for the latter. And I can't even guess what's lost in translation of all those more exotic formats.

To be fair the manual doesn't explicitly say .XM and for the most part .MOD are the formats that are supposed to be fully, losslessly, supported.

I think that adding .IT export just to enable >32 channel support a) would be misleading and b) could still mean a disproportionate amount of work. An existing method of extending the channel range would of course still be preferable to maintain some interoperability. By the sound of OpenMPT's "hack" (as they call it), I'm inclined to think it'd be a more lightweight and feasible solution to the issue.

**Comment by alugarius** [comment]
OK, I understand and im ok with it. 

From my side this issue can be closed, and I suggest to get this to the FAQ

Thanks @SyntaxErrol

**Comment by alugarius** [comment]
Peter send me an email, he will look into it next week.
(Thanks again)

**Comment by alugarius** [comment]
i saw the commit in the testing branch,
id like to do some testing

**Comment by FyiurAmron** [comment]
>If we add more channels, should .mod be used or an own format?
edit: or xm and mod for 64 channels since I found this

If the file is XM-compatible with >32 channels, the extension should probably be simply XM.
If the file is MOD-compatible with >32 channels, the extension should probably be simply MOD.

The main reason is that that's AFAIK the way existing software (XMPlay, BASS, OpenMPT, Sk@le etc.) works.

**Comment by Deltafire** [comment]
@alugarius - you should be able to obtain a Windows build from AppVeyor:

[https://ci.appveyor.com/project/Deltafire/milkytracker](https://ci.appveyor.com/project/Deltafire/milkytracker)

Select a configuration/job, you'll find the binary under the Artifacts tab.

**Comment by alugarius** [comment]
@Deltafire 
I tested with the AppVeyor builds, i cant add more than 32 channels and i cant load the testfile menioned above.

**Comment by Deltafire** [comment]
You have to enable it first, it's a config setting under the IO section (I think), it's not visible on the lower screen resulutions.

**Comment by alugarius** [comment]
thanks

**Comment by FyiurAmron** [comment]
it crashes on my box as soon as 128 ch. is enabled in options (immediately after clicking 'apply'; if I save the config instead, it crashes on app reload). It's probably some trivial error, I'll debug it if I have a spare minute.

64 ch. works perfectly, including scopes.

**Comment by Deltafire** [comment]
Tagging @milkypailes

**Comment by Deltafire** [comment]
Feature added in PR #104
